git_theta/git_utils.py

```diff
@@ -143,60 +178,136 @@ def write_gitattributes(
-        # is escaped due to having spaces in it.
-        match = re.match(r"^\s*(?P<pattern>[^\s]+)\s+(?P<attributes>.*)$", line)
-        if match:
-            # If there is already a pattern that covers the file, add the filter
-            # to that.
-            if fnmatch.fnmatchcase(path, match.group("pattern")):
-                pattern_found = True
-                if not "filter=theta" in match.group("attributes"):
-                    line = f"{line.rstrip()} filter=theta"
-                if not "merge=theta" in match.group("attributes"):
-                    line = f"{line.rstrip()} merge=theta"
-                if not "diff=theta" in match.group("attributes"):
-                    line = f"{line.rstrip()} diff=theta"
-        new_gitattributes.append(line)
-    # If we don't find a matching pattern, add a new line that covers just this
-    # specific file.
-    if not pattern_found:
-        new_gitattributes.append(f"{path} filter=theta merge=theta diff=theta")
-    return new_gitattributes
+def add_theta_to_gitattributes(
+    gitattributes: List[GitAttributes], path: str
+) -> List[GitAttributes]:
+    """Add git attributes required by git-theta for path."""
+    theta_attributes = ("filter", "merge", "diff")
+    previous_attribute = None
+    # Find if an active gitattribute entry applies to path
+    for gitattribute in gitattributes[::-1]:
+        if fnmatch.fnmatchcase(path, gitattribute.pattern):
+            previous_attribute = gitattribute
+            break
+    # If path is already managed by a git attributes entry.
+    if previous_attribute:
+        # If all of the theta attributes are set, we don't do anything.
+        if all(
+            previous_attribute.attributes.get(attr) == "theta"
+            for attr in theta_attributes
+        ):
+            return gitattributes
+        # If any of the attributes theta uses is set to something else, error out.
+        if any(
+            attr in previous_attribute.attributes
+            and previous_attribute.attributes[attr] != "theta"
+            for attr in theta_attributes
+        ):
+            raise ValueError(
+                f"Git Attributes used by git-theta are already set for {path}. "
+                f"Found filter={previous_attribute.attributes.get('filter')}, "
+                f"diff={previous_attribute.attributes.get('diff')}, "
+                f"merge={previous_attribute.attributes.get('merge')}."
+            )
+    # If the old entry set other attributes, make sure they are preserved.
+    attributes = (
+        copy.deepcopy(previous_attribute.attributes) if previous_attribute else {}
+    )
+    for attr in theta_attributes:
+        attributes[attr] = "theta"
+    new_attribute = GitAttributes(path, attributes)
+    gitattributes.append(new_attribute)
+    return gitattributes
+
+
+# def add_theta_to_gitattributes(gitattributes: List[str], path: str) -> str:
```

Comment:
Reminder to remove this commented-out bit.

Reply:
done